### ISSUE TITLE
Fantomas version

### ISSUE THREAD

**Issue by jonekdahl** [JetBrains/resharper-fsharp#234]
Hi,

I understand that fsharp-support bundles Fantomas and I have a few questions regarding the Fantomas version:

 * Can I see the Fantomas version in Rider?
 * If not, which version is bundled with 2020.3?
 * Can I change the Fantomas version used by Rider? (some of our repositories are using older Fantomas versions)

Thanks, Jon

**Comment by auduchinok** [comment]
Hi Jon,

We bundle a custom Fantomas build that is usually forked from roughly the latest version at the time of a Rider release.
Rider 2020.3 uses ~4.2 version, and 2021.1 will base on 4.4.

If you need to override Fantomas version in Rider, it's possible to use it as an [external tool](https://github.com/fsprojects/fantomas#using-the-latest-version-inside-rider), though we don't test this scenario and some bits of the integration may not work as expected.

**Comment by auduchinok** [comment]
Could you tell a bit more about reasons to stay on older Fantomas versions, please? Are there some breaking changes or bugs that prevent you from upgrading? cc @nojaf

**Comment by jonekdahl** [comment]
Hi Eugene,

> Could you tell a bit more about reasons to stay on older Fantomas versions, please?

Well, I have assumed that different versions of Fantomas formats code differently. If upgrading my Rider suddenly formats code differently then _everyone_ who works in the same repository must also upgrade to the same Fantomas version, otherwise we will keep reformatting each others code, causing a lot of noise in git diffs and pull requests.

We can try to mitigate this by communicating a "known good" Fantomas/Rider version internally, but as the number of developers and repositories increase, this coordination becomes unwieldy or even impossible. What we need is a way to configure the Fantomas version, per repository, and have our tools (including Rider) pick that up.

So to summarize, it's not that we want to stay on older versions, it's just that it's hard to switch every developer and repository to a new formatting simultaneously.

**Comment by nojaf** [comment]
Hello Jon, I recognize this problem and what I do in my projects is pin down a version of Fantomas via [a tool manifest](https://docs.microsoft.com/en-us/dotnet/core/tools/local-tools-how-to-use).
Usage in Rider is then indeed limited to configuring it as [an external tool](https://github.com/fsprojects/fantomas#using-the-latest-version-inside-rider).

Every once in a while I update the local Fantomas tool as part of a separate PR ([example](https://github.com/fsprojects/fantomas/commit/33044705e3b0f40b56b278e3d46c58786f5b68d5)).

If you are brave enough you could also configure Fantomas as [pre-commit hook](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#a-git-pre-commit-hook-sample) in git.
Personally, I use a [FAKE target](https://github.com/fsprojects/fantomas/blob/ed7a94071cc7deeba849ec7806d67801094b37e6/build.fsx#L381-L398) to format my changed code.

Perhaps it is also an option Eugene to sync to a tag of Fantomas for the Rider integration?

**Comment by auduchinok** [comment]
@jonekdahl Thanks! It makes sense to me now.

@nojaf The problem here is the custom builds of both FCS and Fantomas that we bundle, and Fantomas won't be able to use SyntaxTree from a different strong-name FCS build. I'll think about having way to use the published Fantomas version if it's somehow specified in a project.

**Comment by jonekdahl** [comment]
@nojaf Thanks, I guess the external tool option is the one we have to go for.

**Comment by auduchinok** [comment]
@jonekdahl We've discussed it briefly with @nojaf and we'll think about having an automated way to use external Fantomas tool in future Rider/Fantomas version.

**Comment by jonekdahl** [comment]
That would be awesome, looking forward to it!

**Comment by jonekdahl** [comment]
@auduchinok Should I close this issue or do you want to keep it open?

**Comment by auduchinok** [comment]
@jonekdahl Let's keep it open for a while as a reminder.

**Comment by l3m** [comment]
It seems that Fantomas is missing from the open-source software list available in Rider in Help->About->Powered by open source software

**Comment by auduchinok** [comment]
@l3m Thanks for letting us know, I've forwarded it to the team (see [RIDER-61672](https://youtrack.jetbrains.com/issue/RIDER-61672)).

**Comment by DedSec256** [comment]
@johnekdahl, can you share how you are currently anchoring the Fantomas version in your organization?

**Comment by auduchinok** [comment]
This is implemented in #339, kudos to @DedSec256 and @Rizzen!
